### PR TITLE
Delete unnecessary null pointer checks

### PR DESCRIPTION
**Pull request by elfring**
[The C++ programming language tolerates the passing of null pointers for delete expressions](https://isocpp.org/wiki/faq/freestore-mgmt#delete-handles-null "Do I need to check for null before delete p?").

* Thus omit extra checks at a few places.
* Remove also the resetting of two variables.
* The source code could be simplified a bit for three iterators.

### CODE REVIEW

File: src/Probabilities.cpp [hunk 1]
```diff
@@ -192,9 +192,7 @@ GroupProbElem::~GroupProbElem()
 {
 	std::map<ProbName, SingleProbElem*>::iterator i;
 	for (i = probs_.begin(); i != probs_.end(); ++i) {
-		SingleProbElem *elem = (*i).second;
-		assert(elem);
```
Comment (regehr) [inline src/Probabilities.cpp]: please refrain from removing assertions

File: src/SimpleDeltaSequence.cpp [hunk 2]
```diff
@@ -184,8 +184,7 @@ SimpleDeltaSequence::clear()
 {
 	std::map<int, SimpleDeltaSequence::ValuePair*>::iterator i;
 	for (i = seq_map_.begin(); i != seq_map_.end(); ++i) {
-		assert((*i).second);
```
Comment (regehr) [inline src/SimpleDeltaSequence.cpp]: also here

File: src/Variable.cpp [hunk 3]
```diff
@@ -474,10 +474,7 @@ Variable::~Variable(void)
 	for(i = field_vars.begin(); i != field_vars.end(); ++i)
 		delete (*i);
 	field_vars.clear();
-	if (init) {
-		delete init;
-		init = NULL;
```
Comment (regehr) [inline src/Variable.cpp]: in a patch like this, we do not want changes that are unrelated to the task at hand, so do not remove assignments like this

File: src/Probabilities.cpp (hunk 1 above)
Comment (elfring) [reply src/Probabilities.cpp]: Why would you insist to keep such a questionable null pointer check (before the usage of a delete expression)?

File: src/Variable.cpp (hunk 3 above)
Comment (elfring) [reply src/Variable.cpp]: Would you like to get another bug report for such an implementation detail?

File: src/Probabilities.cpp (hunk 1 above)
Comment (regehr) [reply src/Probabilities.cpp]: if the application has an invariant, and if a developer felt the need to check this invariant, then a refactoring change should not remove the check. the delete has nothing to do with it.

Comment (regehr) [reply src/Probabilities.cpp]: sorry, I don't have time for this
